### PR TITLE
TAJO-2122: PullServer as an Auxiliary service of Yarn

### PR DESCRIPTION
**Pull request by jihoonson**

### CODE REVIEW

File: tajo-core/src/main/java/org/apache/tajo/worker/TaskImpl.java [hunk 1]
```diff
@@ -696,44 +696,48 @@ private synchronized void fetcherFinished(TaskAttemptContext ctx) {
...
-                localChunk = new FileChunk(defaultStoreFile, 0, -1);
-                localChunk.setFromRemote(true);
-              }
-              localChunk.setEbId(f.getName());
-              storeChunkList.add(localChunk);
-            }
+            localStoreChunkCount++;
+            runnerList.add(new LocalFetcher(systemConf, uri, executionBlockContext, f.getName()));
+
+//            List<FileChunk> localChunkCandidates = getLocalStoredFileChunk(uri, systemConf);
```
Comment (jinossy) [inline tajo-core/src/main/java/org/apache/tajo/worker/TaskImpl.java]: Could you explain why you comment out this block?

File: tajo-core/src/main/java/org/apache/tajo/worker/LocalFetcher.java [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+          .option(ChannelOption.TCP_NODELAY, true);
+    } else {
+      endFetch(FetcherState.FETCH_FAILED);
+      throw new TajoInternalError("Pull server service is not initialized");
+    }
+  }
+
+  @Override
+  public List<FileChunk> get() throws IOException {
+    return pullServerService != null ? getDirect() : getFromFetchURI();
```
Comment (jinossy) [inline tajo-core/src/main/java/org/apache/tajo/worker/LocalFetcher.java]: Why local fetcher connect to remote server?

File: tajo-core/src/main/java/org/apache/tajo/worker/TaskImpl.java (hunk 1 above)
Comment (jihoonson) [inline tajo-core/src/main/java/org/apache/tajo/worker/TaskImpl.java]: This code is not used anymore. I forgot to remove it.

File: tajo-core/src/main/java/org/apache/tajo/worker/LocalFetcher.java (hunk 2 above)
Comment (jihoonson) [inline tajo-core/src/main/java/org/apache/tajo/worker/LocalFetcher.java]: There are two cases for local fetcher. 
* When an internal pull server is running, local fetchers can retrieve data directly. 
* When an external pull server is running, 
 * If the shuffle type is hash, local fetchers can still retrieve data directly.
 * If the shuffle type is range, local fetchers need to get meta information of data via HTTP. Once the meta information is retrieved, they can read data directly. 

I'll add this description.

Comment (jinossy) [inline tajo-core/src/main/java/org/apache/tajo/worker/LocalFetcher.java]: OK, thanks

File: tajo-yarn/pom.xml [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+        </configuration>
+      </plugin>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-surefire-report-plugin</artifactId>
+      </plugin>
+    </plugins>
+  </reporting>
+
+</project>
```
Comment (jinossy) [inline tajo-yarn/pom.xml]: @jihoonson 
There are included too many dependencies. Could exclude unused module ?
```
[INFO] ------------------------------------------------------------------------
[INFO] Building Tajo Yarn 0.12.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO]
[INFO] --- maven-dependency-plugin:2.4:tree (default-cli) @ tajo-yarn ---
[INFO] org.apache.tajo:tajo-yarn:jar:0.12.0-SNAPSHOT
[INFO] +- org.apache.hadoop:hadoop-mapreduce-client-shuffle:jar:2.7.2:provided
[INFO] |  +- org.apache.hadoop:hadoop-yarn-server-common:jar:2.7.2:provided
[INFO] |  |  \- org.apache.hadoop:hadoop-yarn-common:jar:2.7.2:provided
[INFO] |  |     +- javax.xml.bind:jaxb-api:jar:2.2.2:provided
[INFO] |  |     |  +- javax.xml.stream:stax-api:jar:1.0-2:provided
[INFO] |  |     |  \- javax.activation:activation:jar:1.1:provided
[INFO] |  |     +- com.sun.jersey:jersey-client:jar:1.9:provided
[INFO] |  |     \- com.sun.jersey.contribs:jersey-guice:jar:1.9:provided
[INFO] |  +- org.apache.hadoop:hadoop-mapreduce-client-common:jar:2.7.2:provided
[INFO] |  |  \- org.apache.hadoop:hadoop-yarn-client:jar:2.7.2:provided
[INFO] |  +- org.fusesource.leveldbjni:leveldbjni-all:jar:1.8:provided
[INFO] |  +- org.slf4j:slf4j-log4j12:jar:1.7.10:provided
[INFO] |  +- org.apache.hadoop:hadoop-annotations:jar:2.7.2:provided
[INFO] |  |  \- jdk.tools:jdk.tools:jar:1.8:system
[INFO] |  \- com.google.inject.extensions:guice-servlet:jar:3.0:provided
[INFO] |     \- com.google.inject:guice:jar:3.0:provided
[INFO] |        +- javax.inject:javax.inject:jar:1:provided
[INFO] |        \- aopalliance:aopalliance:jar:1.0:provided
[INFO] +- org.apache.hadoop:hadoop-yarn-api:jar:2.7.2:provided
[INFO] +- org.apache.hadoop:hadoop-common:jar:2.7.2:provided (scope not updated to compile)
[INFO] |  +- commons-cli:commons-cli:jar:1.2:provided
[INFO] |  +- org.apache.commons:commons-math3:jar:3.1.1:provided
[INFO] |  +- xmlenc:xmlenc:jar:0.52:provided
[INFO] |  +- commons-httpclient:commons-httpclient:jar:3.1:provided
[INFO] |  +- commons-codec:commons-codec:jar:1.10:provided
[INFO] |  +- commons-io:commons-io:jar:2.4:provided
[INFO] |  +- commons-net:commons-net:jar:3.1:provided
[INFO] |  +- javax.servlet:servlet-api:jar:2.5:provided
[INFO] |  +- javax.servlet.jsp:jsp-api:jar:2.1:provided
[INFO] |  +- com.sun.jersey:jersey-core:jar:1.9:provided
[INFO] |  +- com.sun.jersey:jersey-json:jar:1.9:provided
[INFO] |  |  +- org.codehaus.jettison:jettison:jar:1.1:provided
[INFO] |  |  +- com.sun.xml.bind:jaxb-impl:jar:2.2.3-1:provided
[INFO] |  |  +- org.codehaus.jackson:jackson-jaxrs:jar:1.8.3:provided
[INFO] |  |  \- org.codehaus.jackson:jackson-xc:jar:1.8.3:provided
[INFO] |  +- com.sun.jersey:jersey-server:jar:1.9:provided
[INFO] |  +- log4j:log4j:jar:1.2.17:provided
[INFO] |  +- net.java.dev.jets3t:jets3t:jar:0.9.0:provided
[INFO] |  |  +- org.apache.httpcomponents:httpclient:jar:4.1.2:provided
[INFO] |  |  +- org.apache.httpcomponents:httpcore:jar:4.1.2:provided
[INFO] |  |  \- com.jamesmurty.utils:java-xmlbuilder:jar:0.4:provided
[INFO] |  +- commons-configuration:commons-configuration:jar:1.6:provided
[INFO] |  |  +- commons-digester:commons-digester:jar:1.8:provided
[INFO] |  |  |  \- commons-beanutils:commons-beanutils:jar:1.7.0:provided
[INFO] |  |  \- commons-beanutils:commons-beanutils-core:jar:1.8.0:provided
[INFO] |  +- org.apache.hadoop:hadoop-auth:jar:2.7.2:provided
[INFO] |  |  +- org.apache.directory.server:apacheds-kerberos-codec:jar:2.0.0-M15:provided
[INFO] |  |  |  +- org.apache.directory.server:apacheds-i18n:jar:2.0.0-M15:provided
[INFO] |  |  |  +- org.apache.directory.api:api-asn1-api:jar:1.0.0-M20:provided
[INFO] |  |  |  \- org.apache.directory.api:api-util:jar:1.0.0-M20:provided
[INFO] |  |  \- org.apache.curator:curator-framework:jar:2.7.1:provided
[INFO] |  +- com.jcraft:jsch:jar:0.1.42:provided
[INFO] |  +- org.apache.curator:curator-client:jar:2.7.1:provided
[INFO] |  +- org.apache.curator:curator-recipes:jar:2.7.1:provided
[INFO] |  +- org.apache.htrace:htrace-core:jar:3.1.0-incubating:provided
[INFO] |  +- org.apache.zookeeper:zookeeper:jar:3.4.6:provided
[INFO] |  \- org.apache.commons:commons-compress:jar:1.4.1:provided
[INFO] |     \- org.tukaani:xz:jar:1.0:provided
[INFO] +- org.apache.tajo:tajo-common:jar:0.12.0-SNAPSHOT:compile
[INFO] |  +- com.google.protobuf:protobuf-java:jar:2.5.0:compile
[INFO] |  +- commons-logging:commons-logging:jar:1.1.1:compile
[INFO] |  +- commons-logging:commons-logging-api:jar:1.1:compile
[INFO] |  +- commons-lang:commons-lang:jar:2.6:compile
[INFO] |  +- com.google.guava:guava:jar:11.0.2:compile
[INFO] |  |  \- com.google.code.findbugs:jsr305:jar:3.0.0:compile
[INFO] |  +- com.google.code.gson:gson:jar:2.2.2:compile
[INFO] |  +- io.netty:netty-buffer:jar:4.0.34.Final:compile
[INFO] |  |  \- io.netty:netty-common:jar:4.0.34.Final:compile
[INFO] |  \- org.iq80.snappy:snappy:jar:0.4:compile
[INFO] +- org.apache.tajo:tajo-storage-common:jar:0.12.0-SNAPSHOT:compile
[INFO] |  +- org.apache.tajo:tajo-catalog-common:jar:0.12.0-SNAPSHOT:compile
[INFO] |  +- org.apache.tajo:tajo-plan:jar:0.12.0-SNAPSHOT:compile
[INFO] |  |  \- org.apache.tajo:tajo-algebra:jar:0.12.0-SNAPSHOT:compile
[INFO] |  \- net.minidev:json-smart:jar:2.1.1:compile
[INFO] |     \- net.minidev:asm:jar:1.0.2:compile
[INFO] |        \- asm:asm:jar:3.1:compile
[INFO] +- org.apache.tajo:tajo-storage-hdfs:jar:0.12.0-SNAPSHOT:compile
[INFO] |  +- io.netty:netty-transport:jar:4.0.34.Final:compile
[INFO] |  +- io.netty:netty-codec:jar:4.0.34.Final:compile
[INFO] |  +- io.netty:netty-codec-http:jar:4.0.34.Final:compile
[INFO] |  |  \- io.netty:netty-handler:jar:4.0.34.Final:compile
[INFO] |  +- org.apache.avro:trevni-core:jar:1.7.3:compile
[INFO] |  |  +- org.xerial.snappy:snappy-java:jar:1.0.4.1:compile
[INFO] |  |  \- org.slf4j:slf4j-api:jar:1.7.10:compile
[INFO] |  +- org.apache.avro:trevni-avro:jar:1.7.3:compile
[INFO] |  |  \- org.apache.avro:avro-mapred:jar:1.7.3:compile
[INFO] |  |     +- org.apache.avro:avro-ipc:jar:1.7.3:compile
[INFO] |  |     |  +- org.apache.avro:avro:jar:1.7.4:compile
[INFO] |  |     |  |  \- com.thoughtworks.paranamer:paranamer:jar:2.3:compile
[INFO] |  |     |  +- org.mortbay.jetty:jetty:jar:6.1.26:compile
[INFO] |  |     |  +- org.mortbay.jetty:jetty-util:jar:6.1.26:compile
[INFO] |  |     |  +- io.netty:netty:jar:3.6.2.Final:compile
[INFO] |  |     |  +- org.apache.velocity:velocity:jar:1.7:compile
[INFO] |  |     |  |  \- commons-collections:commons-collections:jar:3.2.2:compile
[INFO] |  |     |  \- org.mortbay.jetty:servlet-api:jar:2.5-20081211:compile
[INFO] |  |     +- org.apache.avro:avro-ipc:jar:tests:1.7.3:compile
[INFO] |  |     +- org.codehaus.jackson:jackson-core-asl:jar:1.8.8:compile
[INFO] |  |     \- org.codehaus.jackson:jackson-mapper-asl:jar:1.8.8:compile
[INFO] |  +- org.apache.parquet:parquet-hadoop-bundle:jar:1.8.1:compile
[INFO] |  +- org.apache.hive:hive-orc:jar:2.0.0:compile
[INFO] |  \- org.apache.hive:hive-storage-api:jar:2.0.0:compile
[INFO] \- org.apache.tajo:tajo-pullserver:jar:0.12.0-SNAPSHOT:compile
[INFO]    \- org.apache.tajo:tajo-rpc-protobuf:jar:0.12.0-SNAPSHOT:compile
[INFO]       \- org.apache.tajo:tajo-rpc-common:jar:0.12.0-SNAPSHOT:compile
```

Comment (jihoonson) [inline tajo-yarn/pom.xml]: Thanks. I've cleaned up the dependency of module yarn.

Comment (jinossy) [inline tajo-yarn/pom.xml]: Thanks for your quick update. I’ll test on cluster

**Comment by jinossy** [comment]
@jihoonson 
I've test on hadoop-2.7.2 and I found following error
```
2016-04-25 13:49:36,698 WARN org.apache.tajo.pullserver.PullServerUtil: Failed to manage OS cache for /data02/tajo/q_1461559624789_0001/output/1/hash-shuffle/30/158
java.lang.NullPointerException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.tajo.pullserver.PullServerUtil.posixFadviseIfPossible(PullServerUtil.java:84)
	at org.apache.tajo.yarn.FadvisedFileRegion.transferSuccessful(FadvisedFileRegion.java:164)
	at org.apache.tajo.yarn.FileCloseListener.operationComplete(FileCloseListener.java:37)
	at org.jboss.netty.channel.DefaultChannelFuture.notifyListener(DefaultChannelFuture.java:427)
	at org.jboss.netty.channel.DefaultChannelFuture.addListener(DefaultChannelFuture.java:145)
	at org.apache.tajo.yarn.TajoPullServerService$PullServer.sendFile(TajoPullServerService.java:533)
	at org.apache.tajo.yarn.TajoPullServerService$PullServer.handleChunkRequest(TajoPullServerService.java:506)
	at org.apache.tajo.yarn.TajoPullServerService$PullServer.messageReceived(TajoPullServerService.java:369)
	at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:560)
	at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:787)
	at org.jboss.netty.handler.stream.ChunkedWriteHandler.handleUpstream(ChunkedWriteHandler.java:142)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:560)
	at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:787)
	at org.jboss.netty.handler.codec.http.HttpChunkAggregator.messageReceived(HttpChunkAggregator.java:148)
	at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:560)
	at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:787)
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:296)
	at org.jboss.netty.handler.codec.frame.FrameDecoder.unfoldAndFireMessageReceived(FrameDecoder.java:459)
	at org.jboss.netty.handler.codec.replay.ReplayingDecoder.callDecode(ReplayingDecoder.java:536)
	at org.jboss.netty.handler.codec.replay.ReplayingDecoder.messageReceived(ReplayingDecoder.java:435)
	at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
	at org.jboss.netty.handler.codec.http.HttpServerCodec.handleUpstream(HttpServerCodec.java:56)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:560)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:555)
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:268)
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:255)
	at org.jboss.netty.channel.socket.nio.NioWorker.read(NioWorker.java:88)
	at org.jboss.netty.channel.socket.nio.AbstractNioWorker.process(AbstractNioWorker.java:107)
	at org.jboss.netty.channel.socket.nio.AbstractNioSelector.run(AbstractNioSelector.java:312)
	at org.jboss.netty.channel.socket.nio.AbstractNioWorker.run(AbstractNioWorker.java:88)
	at org.jboss.netty.channel.socket.nio.NioWorker.run(NioWorker.java:178)
	at org.jboss.netty.util.ThreadRenamingRunnable.run(ThreadRenamingRunnable.java:108)
	at org.jboss.netty.util.internal.DeadLockProofWorker$1.run(DeadLockProofWorker.java:42)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by jihoonson** [comment]
@jinossy, thanks. I've fixed NPE problem.

**Comment by jinossy** [comment]
+1 Great work!
Thanks for your effort!

**Comment by jihoonson** [comment]
@jinossy, thanks for your review!

**Pull request closed without merging**